File: avaldused/client-dotnet/PersonalData.cs
using System;
using System.Text;

/* This library is not Gemalto-specific, although the namespace would lead you to believe otherwise. */
using GemCard;

namespace EidClient
{
    /// <summary>
    /// Class for reading personal data from an ID-card.
    /// </summary>
    class PersonalData
    {
        private static int NUM_RECORDS = 16;
        private static string[] records = new string[NUM_RECORDS];

        /* Commands used to read the personal records. */
        static APDUCommand
            apduSelectMF = new APDUCommand(0, 0xA4, 0, 0x0C, null, 0),
            apduSelectEEEE = new APDUCommand(0, 0xA4, 1, 0x0C, new byte[] { 0xEE, 0xEE }, 2),
            apduSelect5044 = new APDUCommand(0, 0xA4, 2, 0x0C, new byte[] { 0x50, 0x44 }, 2),
            apduReadRecord = new APDUCommand(0, 0xB2, 0, 4, null, 0),
            apduGetResponse = new APDUCommand(0, 0xC0, 0, 0, null, 0);

        const ushort SC_OK = 0x9000;
        const byte SC_PENDING = 0x61;

        /// <summary>
        /// Reads personal records from the currently inserted ID-card into the <tt>records</tt> array.
        /// </summary>
        private static void readRecords()
        {
            CardNative iCard = new CardNative();
            APDUParam apduParam = new APDUParam();
            APDUResponse apduResp;

            string[] readers = iCard.ListReaders();

            /* Change this index to the one where your card is inserted. This is usually 0,
             * unless you have multiple readers. */
            string reader = readers[0];
            iCard.Connect(reader, SHARE.Shared, PROTOCOL.T0);
            Console.Write("Reading personal data from " + reader + "...");

            try
            {
                /* Select the Master File */
                apduResp = iCard.Transmit(apduSelectMF);
                if (apduResp.Status != SC_OK)
                    throw new Exception("Select MF failed: " + apduResp.ToString());

                /* Select EEEE */
                apduResp = iCard.Transmit(apduSelectEEEE);
                if (apduResp.Status != SC_OK)
                    throw new Exception("Select EEEE failed: " + apduResp.ToString());

                /* Select 5044 */
                apduResp = iCard.Transmit(apduSelect5044);
                if (apduResp.Status != SC_OK)
                    throw new Exception("Select 5044 failed: " + apduResp.ToString());

                /* Read the records */
                for (byte i = 1; i <= NUM_RECORDS; i++)
                {
                    /* Send the read command */
                    apduParam.Reset();
                    apduParam.P1 = i;
                    apduReadRecord.Update(apduParam);
                    apduResp = iCard.Transmit(apduReadRecord);
                    if (apduResp.SW1 != SC_PENDING)
                        throw new Exception("Read record failed: " + apduResp.ToString());

                    /* Read the response */
                    apduParam.Reset();
                    apduParam.Le = apduResp.SW2;
                    apduGetResponse.Update(apduParam);
                    apduResp = iCard.Transmit(apduGetResponse);
                    if (apduResp.Status != SC_OK)
                        throw new Exception("Get response failed: " + apduResp.ToString());
                    records[i - 1] = Encoding.UTF7.GetString(apduResp.Data).Trim();
                }
            }
            finally
            {
                iCard.Disconnect(DISCONNECT.Leave);
            }
        }

        /// <summary>
        /// Read personal records from the card and return a formatted string.
        /// </summary>
        /// <returns>Personal records on the currently inserted ID-card.</returns>
        public static new string ToString()
        {
            readRecords();
            return "Surname: " + records[0] + '\n'
                + "Given name 1: " + records[1] + '\n'
                + "Given name 2: " + records[2] + '\n'
                + "Gender: " + records[3] + '\n'
                + "Citizenship: " + records[4] + '\n'
                + "Date of birth: " + records[5] + '\n'
                + "Personal code: " + records[6] + '\n'
                + "Document number: " + records[7] + '\n'
                + "Date of expiry: " + records[8] + '\n'
                + "Place of birth: " + records[9] + '\n'
                + "Date of issue: " + records[10] + '\n'
                + "Residence permit type: " + records[11] + '\n'
                + "Remark 1: " + records[12] + '\n'
                + "Remark 2: " + records[13] + '\n'
                + "Remark 3: " + records[14] + '\n'
                + "Remark 4: " + records[15];
        }
    }
}

File: avaldused/client-java/ee/cyber/eid/card/PersonalData.java
package ee.cyber.eid.card;

import java.io.UnsupportedEncodingException;

import javax.smartcardio.CardChannel;
import javax.smartcardio.CardException;
import javax.smartcardio.CommandAPDU;

import ee.cyber.eid.EidException;

/** The personal data stored on an ID-card. */
public class PersonalData {

    /** The number of personal data fields. */
    private static final int RECORDS_COUNT = 16;

    /** The command to select the file 5044. */
    private static final CommandAPDU SELECT_FILE_5044 = new CommandAPDU(
            new byte[] { 0x00, (byte) 0xa4, 0x02, 0x04, 0x02, 0x50, 0x44 });

    private String[] records = new String[RECORDS_COUNT];

    /** Read personal data from a card using the given channel. */
    public PersonalData(CardChannel channel) throws CardException,
            EidException, UnsupportedEncodingException {
        /* Navigate to the personal data file. */
        CardUtil.sendCommand(channel, EstEID.SELECT_FILE_MF);
        CardUtil.sendCommand(channel, EstEID.SELECT_FILE_EEEE);
        CardUtil.sendCommand(channel, SELECT_FILE_5044);

        /* Read the stored personal data into an array. */
        for (byte i = 1; i <= RECORDS_COUNT; i++) {
            records[i - 1] = new String(readRecord(channel, i),
                    CardUtil.CARD_ENCODING).trim();
        }
    }

    /** Read a record from the card using the given channel. */
    private byte[] readRecord(CardChannel channel, byte recordNumber)
            throws CardException, EidException {
        return CardUtil.sendCommand(channel, new CommandAPDU(new byte[] {
                0x00, (byte) 0xb2, recordNumber, 0x04, 0x00 }));
    }

    public String getSurname() {
        return records[0];
    }

    public String getGivenName1() {
        return records[1];
    }

    public String getGivenName2() {
        return records[2];
    }

    public String getGender() {
        return records[3];
    }

    public String getCitizenship() {
        return records[4];
    }

    public String getDateOfBirth() {
        return records[5];
    }

    public String getPersonalCode() {
        return records[6];
    }

    public String getDocumentNumber() {
        return records[7];
    }

    public String getDateOfExpiry() {
        return records[8];
    }

    public String getPlaceOfBirth() {
        return records[9];
    }

    public String getDateOfIssue() {
        return records[10];
    }

    public String getResidencePermitType() {
        return records[11];
    }

    public String getRemark1() {
        return records[12];
    }

    public String getRemark2() {
        return records[13];
    }

    public String getRemark3() {
        return records[14];
    }

    public String getRemark4() {
        return records[15];
    }

    @Override
    public String toString() {
        return "Surname: " + getSurname() + '\n'
                + "Given name 1: " + getGivenName1() + '\n'
                + "Given name 2: " + getGivenName2() + '\n'
                + "Gender: " + getGender() + '\n'
                + "Citizenship: " + getCitizenship() + '\n'
                + "Date of birth: " + getDateOfBirth() + '\n'
                + "Personal code: " + getPersonalCode() + '\n'
                + "Document number: " + getDocumentNumber() + '\n'
                + "Date of expiry: " + getDateOfExpiry() + '\n'
                + "Place of birth: " + getPlaceOfBirth() + '\n'
                + "Date of issue: " + getDateOfIssue() + '\n'
                + "Residence permit type: " + getResidencePermitType() + '\n'
                + "Remark 1: " + getRemark1() + '\n'
                + "Remark 2: " + getRemark2() + '\n'
                + "Remark 3: " + getRemark3() + '\n'
                + "Remark 4: " + getRemark4();
    }

}
